fix: Original object not retrieved when user logs in with third party auth

Meglali20 opened the pull request:
### New Pull Request Checklist
<!--
    Please check the following boxes [x] before submitting your issue.
    Click the "Preview" tab for better readability.
    Thanks for contributing to Parse Server!
-->

- [x] I am not disclosing a [vulnerability](https://github.com/parse-community/parse-server/blob/master/SECURITY.md).
- [x] I am creating this PR in reference to an [issue](https://github.com/parse-community/parse-server/issues?q=is%3Aissue).

### Issue Description
<!-- Add a brief description of the issue this PR solves. -->

Related issue: [issue#7737](https://github.com/parse-community/parse-server/issues/7737)

### Approach
<!-- Add a description of the approach in this PR. -->

### TODOs before merging
<!--
    Add TODOs that need to be completed before merging this PR.
    Delete suggested TODOs that do not apply to this PR.
-->

- [ ] Add tests
- [ ] Add changes to documentation (guides, repository pages, in-code descriptions)
- [ ] Add [security check](https://github.com/parse-community/parse-server/blob/master/CONTRIBUTING.md#security-checks)
- [ ] Add new Parse Error codes to Parse JS SDK <!-- no hard-coded error codes in Parse Server -->
- [x] A changelog entry is created automatically using the pull request title (do not manually add a changelog entry)

Meglali20 committed:
Update ParseUser.spec.js

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -13,6 +13,91 @@ const passwordCrypto = require('../lib/password');
 const Config = require('../lib/Config');
 const cryptoUtils = require('../lib/cryptoUtils');
 
+describe('Parse.User auth after save testing (original object not being retrieved) ', () => {
```
Could you remove this new `describe` and move the `it` under the existing  `Parse.User testing'`?

And could you rename the test to `it('retrieves original object when user logs in with third party auth',`. I think that sums up the issue, right?

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -13,6 +13,91 @@ const passwordCrypto = require('../lib/password');
 
+describe('Parse.User auth after save testing (original object not being retrieved) ', () => {
+  it('check if user is new when using signup or edit with auth login methods', async done => {
+     let counter = 0;
```
Please run lint on the test you added to fix the lint issues.

Meglali20 replied:
is it okay now?

Meglali20 committed:
retrieves original object when user logs in with third party auth

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -13,92 +13,40 @@ const passwordCrypto = require('../lib/password');
+    Parse.Cloud.beforeSave(Parse.User, async request => {
+      const { object: user } = request;
+      if (user.get('authData') != null && user.isNew()) {
+        console.log('Before save: new user');
```
Could you remove all console logs from the test and maybe simplify the test by making it more compact and removing unnecessary code.

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -13,92 +13,40 @@ const passwordCrypto = require('../lib/password');
+    const provider = getMockFacebookProvider();
+    Parse.User._registerAuthenticationProvider(provider);
+    await Parse.User.logOut();
+    await Parse.User._logInWith('facebook');
```
`_logInWith` is a private method and [deprecated](http://parseplatform.org/Parse-SDK-JS/api/3.4.0/Parse.User.html#._logInWith), why did you not use [linkWith](https://docs.parseplatform.org/js/guide/#signing-up-and-logging-in)?

Meglali20 replied:
I did use it because I saw that it's the one being used in most of the tests and also because the purpose of the test is just to demonstrate where the issue is, the issue occurs even when you use the android sdk, also the link method is used to link an already registered account, in this case, I have tested it only when a user performs a signup/login. even if I use what you proposed I get the error 
`Unhandled promise rejection: TypeError: Parse.User._linkWith is not a function`

mtrezza replied:
> I get the error `Unhandled promise rejection: TypeError: Parse.User._linkWith is not a function`

Can you try `Parse.User.linkWith`?

> the link method is used to link an already registered account

It should be for both, new sign up and login with existing account.

> the issue occurs even when you use the android sdk

Let's try to reproduce it without Android SDK, so we know it's a server issue not an SDK issue.

Meglali20 replied:
same error for linkWith
TypeError: Parse.User.linkWith is not a function

Meglali20 committed:
removed unnecessary code

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -14,6 +14,24 @@ const Config = require('../lib/Config');
+        const user = request.object;
+        if (user.get('authData') != null) {
+          if (counter > 0) {
+            done.fail();
```
Can you simplify this code?

Form your issue description I understand that we should expect 2 things when the existing user logs in:
- that authData is not erased
- that the ParseUser.isNew is false

instead of `done.fail` you can use `expect(user.get('authData')).not.toBeUndefined();` and so on, because this tree of conditions and counter is somewhat difficult of understand and error prone.

Do you know which field is modified that triggers the `afterSave` when an existing user logs in? Because if the `authData` is already set, I am not sure which field would be changed at all.

Meglali20 replied:
The issue isn't only with the Android SDK, did try it also with the JS SDK (facebook login) same issue.
I get also the same error when using linkWith
`Unhandled promise rejection: TypeError: Parse.User.linkWith is not a function`
Can't tell which field is modified, i have tested it multiple times, what i noticed is only that sometimes the authData get erased and it doesn't retrieve the original object in afterSave, also note that the beforeSave is working correctly because it can detect when the user is new.

mtrezza committed:
Merge branch 'alpha' into aftersave-auth-user

mtrezza committed:
ci: fix node engine check (#7891)

dblythy committed:
fix: return correct response when revert is used in beforeSave (#7839)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.1 [skip ci]

## [5.2.1-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.0...5.2.1-alpha.1) (2022-03-26)

### Bug Fixes

* return correct response when revert is used in beforeSave ([#7839](https://github.com/parse-community/parse-server/issues/7839)) ([f63fb2b](https://github.com/parse-community/parse-server/commit/f63fb2b338c908f0e7a648d338c26b9daa50c8f2))

dblythy committed:
perf: reduce database operations when using the constant parameter in Cloud Function validation (#7892)

semantic-release-bot committed:
chore(release): 5.2.1-alpha.2 [skip ci]

## [5.2.1-alpha.2](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.1...5.2.1-alpha.2) (2022-03-26)

### Performance Improvements

* reduce database operations when using the constant parameter in Cloud Function validation ([#7892](https://github.com/parse-community/parse-server/issues/7892)) ([48bd512](https://github.com/parse-community/parse-server/commit/48bd512eeb47666967dff8c5e723ddc5b7801daa))

github-actions[bot] committed:
feat: add MongoDB 5.1 compatibility (#7682)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.1 [skip ci]

# [5.3.0-alpha.1](https://github.com/parse-community/parse-server/compare/5.2.1-alpha.2...5.3.0-alpha.1) (2022-03-27)

### Features

* add MongoDB 5.1 compatibility ([#7682](https://github.com/parse-community/parse-server/issues/7682)) ([90155cf](https://github.com/parse-community/parse-server/commit/90155cf1680e5e0499b0000e071c6cb0ce3aef96))

mtrezza committed:
fix: security upgrade parse push adapter from 4.1.0 to 4.1.2 (#7893)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.2 [skip ci]

# [5.3.0-alpha.2](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.1...5.3.0-alpha.2) (2022-03-27)

### Bug Fixes

* security upgrade parse push adapter from 4.1.0 to 4.1.2 ([#7893](https://github.com/parse-community/parse-server/issues/7893)) ([ef56e98](https://github.com/parse-community/parse-server/commit/ef56e98ef65041b4d3b7b82cce3473269c27f6fd))

mtrezza committed:
feat: add MongoDB 5.2 support (#7894)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.3 [skip ci]

# [5.3.0-alpha.3](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.2...5.3.0-alpha.3) (2022-03-27)

### Features

* add MongoDB 5.2 support ([#7894](https://github.com/parse-community/parse-server/issues/7894)) ([6b4b358](https://github.com/parse-community/parse-server/commit/6b4b358f0842ae920e45652f5e8b2afebc6caf3a))

snyk-bot committed:
refactor: upgrade mongodb from 4.3.1 to 4.4.1 (#7906)

github-actions[bot] committed:
ci: bump environment (#7907)

Moumouls committed:
refactor: add missing schema definitions (#7917)

dblythy committed:
fix: custom database options are not passed to MongoDB GridFS (#7911)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.4 [skip ci]

# [5.3.0-alpha.4](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.3...5.3.0-alpha.4) (2022-04-04)

### Bug Fixes

* custom database options are not passed to MongoDB GridFS ([#7911](https://github.com/parse-community/parse-server/issues/7911)) ([a72b384](https://github.com/parse-community/parse-server/commit/a72b384f76137a3d83ffb69f65cb25aff1bbab4f))

davimacedo reviewed (commented):
I believe this test case will never succeed and always timeout because there is no call to the `done()` function. Also, even to reproduce the related issue, I think we should have something like:

```
await Parse.User._logInWith('facebook');
await Parse.User.logOut();
await Parse.User._logInWith('facebook');
```

I mean, calling the `_loginWith` twice. But, doing that, I'm afraid that Parse Server will possibly always identify the user as a new user, because it is not being sent e-mail or username to uniquely identify them.

Finally, could you please also rebase the PR against main branch, because we recently had a huge PR merged regarding auth adapters. So maybe your issues was already fixed.

Meglali20 committed:
Merge branch 'parse-community:alpha' into aftersave-auth-user

Meglali20 commented:
@davimacedo how would you suggest using the same account to log in twice?

davimacedo commented:
@Meglali20 I've just checked `getMockFacebookProvider` and it actually returns same account every time, so you are good with that. It would be good if you could login the user twice and have a success flow calling `done()` function in the test though.

Meglali20 commented:
@davimacedo why a success flow on login? that would work but the purpose is to retrieve the original user to check if the user is new or just performing a login. actually, when I ran the test I had no issue and the test demonstrated that it fails and it isn't retrieving the original object on the second login, and the auth data is there.

davimacedo commented:
When we write a test for a use-case, we should have a fail path (to check the behavior that we did not expect to happen) and a success path (to check the behavior that we expect to happen). In the way you wrote your test, it will always fail (even if the supposed bug is later on fixed), because there is no call to the `done()` function. Therefore, we cannot be sure that the test is failing because of a bug that is actually happening or if it is because there is no success path. I'd expect to have a test case that succeeds when `Parse.User._logInWith('facebook')` is called a single time, and fails when `Parse.User._logInWith('facebook')` is called twice. Maybe you will have to use `getMockFacebookProviderWithIdToken()` instead of `getMockFacebookProvider
()` to pass an account that was not used by the other tests. It would be good to also have some message in here `done.fail()` so we can also easily see that the test is failing when hitting this line.

I've seen you rebases with latest alpha. Is the problem still happening?

Meglali20 commented:
@davimacedo yes using now version 5.3.0-alpha.4 and the problem is still there.

mtrezza commented:
I recently removed many `done()` from jasmine tests (can't remember which repo), because that was indicated to be a deprecated syntax. Instead of using `done()`, the test should just return void and not fail. If a fail is expected, use `expect.toThrow()` or `expectAsync.teBeRejectedWith()`.

Meglali20 committed:
retrieves original object when user logs in with third party auth

Meglali20 commented:
@mtrezza can you check the test again?

snyk-bot committed:
refactor: upgrade @apollo/client from 3.5.8 to 3.5.9 (#7921)

dependabot[bot] committed:
fix: security upgrade moment from 2.29.1 to 2.29.2 (#7931)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.5 [skip ci]

# [5.3.0-alpha.5](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.4...5.3.0-alpha.5) (2022-04-09)

### Bug Fixes

* security upgrade moment from 2.29.1 to 2.29.2 ([#7931](https://github.com/parse-community/parse-server/issues/7931)) ([6b68593](https://github.com/parse-community/parse-server/commit/6b68593eaec17e8b183899d2b92699c9ede7625b))

mtrezza committed:
Merge branch 'alpha' into aftersave-auth-user

marvinroger committed:
fix: peer dependency mismatch for GraphQL dependencies (#7934)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.6 [skip ci]

# [5.3.0-alpha.6](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.5...5.3.0-alpha.6) (2022-04-11)

### Bug Fixes

* peer dependency mismatch for GraphQL dependencies ([#7934](https://github.com/parse-community/parse-server/issues/7934)) ([b7a1d76](https://github.com/parse-community/parse-server/commit/b7a1d7617b4bcac677cecedfeb6ac4a27447083b))

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.2 to 8.2.4 (#7932)

mtrezza committed:
Merge branch 'alpha' into aftersave-auth-user

mtrezza commented:
@Meglali20 It seems the tests are still failing...

Meglali20 commented:
@mtrezza what do you mean by failing? Do you mean the test fails to run? because for me the test runs and shows that its fails and if I change it so that it passes I get no issue

mtrezza commented:
If you take a look at the tests here, you can see that 16 of them are failing.

Meglali20 commented:
@mtrezza when I opened the issue and explained it you asked me for a failing test, the test is provided, and when I run `npm test spec/ParseUser.spec.js` from my side all the test passes except the one that I added fails as intended, also here on GitHub I can see that it fails and that there are some that are still pending. I really didn't get what is the issue, can you please explain it to me? I also just changed _logInWith to logInWith

Meglali20 committed:
_logInWith to logInWith

mtrezza commented on spec/ParseUser.spec.js:
```diff
@@ -14,6 +14,27 @@ const Config = require('../lib/Config');
+    Parse.Cloud.afterSave(Parse.User, async request => {
+      if (!request.original) {
+        const user = request.object;
+        expect(user.get('authData')).not.toBeUndefined();
```
If the afterSave trigger is not executed, then the "expects" inside won't apply. You could use a var outside the trigger and move the "expects" outside, or expect that the trigger has been called.

For example:

```
let user;

Parse.Cloud.afterSave(...) {
  user = request.object;
}

expect(user).toBeDefined();
...
```

Meglali20 replied:
afterSave is supposed to be triggered when logged in with Facebook to update the access token, but I changed the code as you requested.

mtrezza replied:
Can you remove the remaining expect out of the trigger and remove the counter variable, I don't think it's necessary?

Meglali20 replied:
The expect is to confirm that the login is performed by an auth provider, and the counter is necessary it's the one that shows that the issue exists, if you check the issue again I meant that when a first-time user signs up we can't retrieve the original object because he is a new user but each time the user logs in after that the original object should be found but instead it's not found just like it's his first time.

mtrezza replied:
Could you simplify it to this? An `expect` should never be inside a trigger, because it may not get called and the test passes, short circuiting itself.

```
it('retrieves original object when user logs in with third party auth', async () => {
     const provider = getMockFacebookProvider();
     Parse.User._registerAuthenticationProvider(provider);
     await Parse.User.logInWith('facebook');
     await Parse.User.logOut();

     let original;
     let user;
     Parse.Cloud.afterSave(Parse.User, async request => {
       original = request.original;
       user = request.object;
     });
     await Parse.User.logInWith('facebook');
     expect(original).toBeDefined();
     expect(user.get('authData')).not.toBeUndefined();
   });
```

Meglali20 replied:
Done, I just put afterSave before because if you put as you suggested it would be triggered only once, so it has to be before so it gets triggered twice as intended (first should be for signup and second for login).

mtrezza reviewed (changes requested):
Got it, this is only the failing test, I assumed it contains the fix as well.

Meglali20 committed:
move expect to outside afterSave

snyk-bot committed:
refactor: upgrade @graphql-tools/links from 8.2.4 to 8.2.6 (#7935)

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.6 to 8.2.7 (#7941)

dplewis committed:
refactor: upgrade body-parser from 1.19.1 to 1.19.2 (#7944)

dplewis committed:
refactor: upgrade @graphql-tools/links from 8.2.7 to 8.2.8 (#7945)

snyk-bot committed:
fix: security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 (#7948)

semantic-release-bot committed:
chore(release): 5.3.0-alpha.7 [skip ci]

# [5.3.0-alpha.7](https://github.com/parse-community/parse-server/compare/5.3.0-alpha.6...5.3.0-alpha.7) (2022-04-25)

### Bug Fixes

* security upgrade @parse/fs-files-adapter from 1.2.1 to 1.2.2 ([#7948](https://github.com/parse-community/parse-server/issues/7948)) ([20fc4e2](https://github.com/parse-community/parse-server/commit/20fc4e23b53c91aac657f894bd70d049b7525c37))

Meglali20 committed:
Simplify

Meglali20 committed:
Simplify

Meglali20 committed:
Simplify

mtrezza commented:
Thanks for simplifying the test, it looks good to me. Do you have a suggestion for a fix?

mtrezza committed:
Merge branch 'alpha' into aftersave-auth-user

Meglali20 commented:
Unfortunately, I have no idea about it since it hasn't been long since I discovered and started using parse so I am not completely familiar with it and didn't look at the code yet. There could be some workaround to avoid this issue but that would be just adding unnecessary extra data to the database. you should be updating the label of the issue so that maybe someone can take a look at it and propose a solution.

Meglali20 commented:
@mtrezza any update?

mtrezza commented:
@parse-community/server anyone has a suggestion for a fix? The failing test here already demos the issue.

dblythy commented:
I would assume the issue is happening somewhere around here

https://github.com/parse-community/parse-server/blob/c1e808f9e807fc49508acbde0d8b3f2b901a1638/src/RestWrite.js#L1604

mtrezza commented:
@Meglali20 Would you want to rebase this PR on alpha and resolve the conflicts?

Meglali20 commented:
@mtrezza you mean if I could fix it? I didn't have time yet to look at the parse server code, so I don't think I could fix it at the moment, If you could provide me with some information on how the process works I would maybe try to figure it out.

mtrezza commented:
@Meglali20 I didn't mean to fix the issue, just to fix the conflicts in this PR:
![image](https://user-images.githubusercontent.com/5673677/173060660-954d862c-5c87-4961-96e0-bdd036ac5c8f.png)


There have been commits to `alpha` since you opened this PR, so there are some conflicts between the latest code and this PR.

Meglali20 committed:
Merge branch 'alpha' into aftersave-auth-user

Meglali20 commented:
Resolved conflicts.

mtrezza commented on src/RestWrite.js:
```diff
@@ -1712,4 +1712,4 @@ RestWrite.prototype._updateResponseWithData = function (response, data) {
 
 export default RestWrite;
-module.exports = RestWrite;
+module.exports = RestWrite;
```
```suggestion
module.exports = RestWrite;

```

mtrezza commented on spec/CloudCode.spec.js:
```diff
@@ -3759,4 +3759,4 @@ describe('sendEmail', () => {
     );
   });
-});
+});
```
```suggestion
});

```

mtrezza commented on README.md:
```diff
@@ -1189,4 +1189,4 @@ As of April 5, 2017, Parse, LLC has transferred this code to the parse-community
 [license-svg]: https://img.shields.io/badge/license-BSD-lightgrey.svg
 [license-link]: LICENSE
-[open-collective-link]: https://opencollective.com/parse-server
+[open-collective-link]: https://opencollective.com/parse-server
```
```suggestion
[open-collective-link]: https://opencollective.com/parse-server

```

mtrezza commented on changelogs/CHANGELOG_alpha.md:
```diff
@@ -420,4 +420,4 @@
 - refactor: deprecate `Parse.Cloud.httpRequest`; it is recommended to use a HTTP library instead. (Daniel Blyth) [#7595](https://github.com/parse-community/parse-server/pull/7595)
 - refactor: Modernize HTTPRequest tests (brandongregoryscott) [#7604](https://github.com/parse-community/parse-server/pull/7604)
-- Allow liveQuery on Session class (Daniel Blyth) [#7554](https://github.com/parse-community/parse-server/pull/7554)
+- Allow liveQuery on Session class (Daniel Blyth) [#7554](https://github.com/parse-community/parse-server/pull/7554)
```
```suggestion
- Allow liveQuery on Session class (Daniel Blyth) [#7554](https://github.com/parse-community/parse-server/pull/7554)

```

mtrezza commented on package-lock.json:
```diff
@@ -5914,6 +5914,28 @@
         "vary": "~1.1.2"
       },
       "dependencies": {
+        "body-parser": {
```
Could you revert this file? There have been no package changes

Meglali20 replied:
good now?

Meglali20 replied:
Is it good now?

mtrezza replied:
you can click on the "Files changes" tab at the top; there shouldn't be any changes in the package-lock file

Meglali20 replied:
should be good now.

mtrezza replied:
yes, thanks

mtrezza committed:
Update spec/CloudCode.spec.js

mtrezza committed:
Update README.md

mtrezza committed:
Update src/RestWrite.js

mtrezza committed:
Update changelogs/CHANGELOG_alpha.md

Meglali20 committed:
Update package-lock.json

Meglali20 committed:
Revert "Update package-lock.json"

This reverts commit 76708dc7b8c87d656a31e3a4bcd33150ec0499ec.

Meglali20 committed:
Update package-lock.json

mtrezza commented:
You got it, looks good now

Meglali20 committed:
Merge branch 'alpha' into aftersave-auth-user

JesperLekland commented:
Was there ever a fix for this? Or still just a test that showcases the bug?

Meglali20 commented:
@JesperLekland I am not aware of any fix, I just used a workaround by creating a new class for already registered users.

mtrezza committed:
Merge branch 'alpha' into aftersave-auth-user
